controllers/cluster_controller_test.go

```diff
@@ -404,201 +275,169 @@ func TestUpgradeOSVersion(t *testing.T) {
 		},
 		ObjectMeta: v1.ObjectMeta{
 			Namespace: "default",
-			Name:      "np01",
-			Labels: map[string]string{
-				CAPIWatchFilterLabel:  "v0.4.10",
-				capi.ClusterLabelName: "my-cluster",
-			},
+			Name:      name,
+			Labels:    map[string]string{capi.ClusterLabelName: cluster},
 		},
-		Spec: capzexp.AzureMachinePoolSpec{
-			Template: capzexp.AzureMachineTemplate{
-				Image: &capz.Image{
-					Marketplace: &capz.AzureMarketplaceImage{
-						Publisher: "cncf-upstream",
-						Offer:     "capi",
-						SKU:       "k8s-1dot18dot2-ubuntu-1804",
-						Version:   "latest",
-					},
+		Spec: capz.AzureMachineTemplateSpec{
+			Template: capz.AzureMachineTemplateResource{
+				Spec: capz.AzureMachineSpec{
+					Image: newAzureCapiImage(),
 				},
 			},
 		},
 	}
+}
 
-	machinePool1 := &capiexp.MachinePool{
-		ObjectMeta: v1.ObjectMeta{
-			Namespace: "default",
-			Name:      "np01",
-			Labels: map[string]string{
-				CAPIWatchFilterLabel:  "v0.3.10",
-				capi.ClusterLabelName: "my-cluster",
-			},
-		},
-		Spec: capiexp.MachinePoolSpec{
-			ClusterName: cluster.Name,
-			Template: capi.MachineTemplateSpec{
-				ObjectMeta: capi.ObjectMeta{},
-				Spec: capi.MachineSpec{
-					ClusterName: cluster.Name,
-					InfrastructureRef: v12.ObjectReference{
-						Kind:       azureMachinePool1.Kind,
-						Name:       azureMachinePool1.Name,
-						APIVersion: azureMachinePool1.APIVersion,
-					},
-					Version: to.StringPtr("v1.18.2"),
+func newAzureClusterWithControlPlane() (*capi.Cluster, *kcp.KubeadmControlPlane, *capz.AzureCluster, *capz.AzureMachineTemplate) {
+	cluster, kcp := newClusterWithControlPlane()
+	azureCluster := newAzureCluster(cluster.Name)
+	azureMachineTemplate := newAzureMachineTemplate(cluster.Name)
+
+	cluster.Spec.InfrastructureRef = &v12.ObjectReference{
+		Kind:       azureCluster.Kind,
+		Namespace:  azureCluster.Namespace,
+		Name:       azureCluster.Name,
+		APIVersion: azureCluster.APIVersion,
+	}
+	kcp.Spec.InfrastructureTemplate = v12.ObjectReference{
+		Kind:       azureMachineTemplate.Kind,
+		Namespace:  azureMachineTemplate.Namespace,
+		Name:       azureMachineTemplate.Name,
+		APIVersion: azureMachineTemplate.APIVersion,
+	}
+	kcp.Spec.KubeadmConfigSpec.Files = append(
+		kcp.Spec.KubeadmConfigSpec.Files,
+		v1alpha3.File{
+			ContentFrom: &v1alpha3.FileSource{
+				Secret: v1alpha3.SecretFileSource{
+					Name: fmt.Sprintf("%s-azure-json", azureMachineTemplate.Name),
 				},
 			},
 		},
-		Status: capiexp.MachinePoolStatus{
-			Conditions: capi.Conditions{capi.Condition{
-				Type:   capi.ReadyCondition,
-				Status: v12.ConditionTrue,
-			}},
-		},
-	}
+	)
+
+	return cluster, kcp, azureCluster, azureMachineTemplate
+}
 
-	azureMachinePool2 := &capzexp.AzureMachinePool{
+func newAzureMachinePool(cluster, name string) *capzexp.AzureMachinePool {
+	return &capzexp.AzureMachinePool{
+		// TODO (mig4): I think specifying TypeMeta explicitly is only necessary because
```

Comment:
Actually for our unit tests we don't start the manager at all. Not sure if this can be fixed for unit tests where we won't use the `testenv` thingy.

Reply:
Yeah, yeah, that's what I meant, I don't think it's needed when using testenv or envtest or whatever they call it 😄

Reply:
Anyway, I'll be poking around that a bit more anyway so I'll deal with that comment in another PR.